### ISSUE TITLE
Switching between container runtimes on existing cluster breaks aws-node pods

### ISSUE THREAD

**Issue by ValeriiVozniuk** [kubernetes/kops#11798]
/kind bug

**1. What `kops` version are you running? The command `kops version`, will display
 this information.**
Version 1.19.2 (git-e288df46e173ba8ce44ac52502283206c0d211ee)

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.10", GitCommit:"98d5dc5d36d34a7ee13368a7893dcb400ec4e566", GitTreeState:"clean", BuildDate:"2021-04-15T03:28:42Z", GoVersion:"go1.15.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.10", GitCommit:"98d5dc5d36d34a7ee13368a7893dcb400ec4e566", GitTreeState:"clean", BuildDate:"2021-04-15T03:20:25Z", GoVersion:"go1.15.10", Compiler:"gc", Platform:"linux/amd64"}

**3. What cloud provider are you using?**
AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**
Switching containerRuntime in cluster manifest from docker to containerd (same happens in reverse change)

**5. What happened after the commands executed?**
On some of the running nodes (2 out of 16) aws-node pods were recreated with an image for a new runtime, making those nodes "unhealthy", thus blocking `kops rolling-update cluster` from execution, as cluster looks to be unhealthy

**6. What did you expect to happen?**
aws-node pods on existing nodes would not be replaced, allowing to upgrade cluster with `kops rolling-update cluster`

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: "2018-11-23T19:42:31Z"
  generation: 1
  name: 208.k8s
spec:
  additionalPolicies:
    master: |
      [
        {
          "Effect": "Allow",
          "Action": [
            "acm:DescribeCertificate",
            "acm:ListCertificates",
            "acm:GetCertificate"
          ],
          "Resource": "*"
        },
        {
          "Effect": "Allow",
          "Action": [
            "ec2:AuthorizeSecurityGroupIngress",
            "ec2:CreateSecurityGroup",
            "ec2:CreateTags",
            "ec2:DeleteTags",
            "ec2:DeleteSecurityGroup",
            "ec2:DescribeAccountAttributes",
            "ec2:DescribeAddresses",
            "ec2:DescribeInstances",
            "ec2:DescribeInstanceStatus",
            "ec2:DescribeInternetGateways",
            "ec2:DescribeNetworkInterfaces",
            "ec2:DescribeSecurityGroups",
            "ec2:DescribeSubnets",
            "ec2:DescribeTags",
            "ec2:DescribeVpcs",
            "ec2:ModifyInstanceAttribute",
            "ec2:ModifyNetworkInterfaceAttribute",
            "ec2:RevokeSecurityGroupIngress"
          ],
          "Resource": "*"
        },
        {
          "Effect": "Allow",
          "Action": [
            "elasticloadbalancing:AddListenerCertificates",
            "elasticloadbalancing:AddTags",
            "elasticloadbalancing:CreateListener",
            "elasticloadbalancing:CreateLoadBalancer",
            "elasticloadbalancing:CreateRule",
            "elasticloadbalancing:CreateTargetGroup",
            "elasticloadbalancing:DeleteListener",
            "elasticloadbalancing:DeleteLoadBalancer",
            "elasticloadbalancing:DeleteRule",
            "elasticloadbalancing:DeleteTargetGroup",
            "elasticloadbalancing:DeregisterTargets",
            "elasticloadbalancing:DescribeListenerCertificates",
            "elasticloadbalancing:DescribeListeners",
            "elasticloadbalancing:DescribeLoadBalancers",
            "elasticloadbalancing:DescribeLoadBalancerAttributes",
            "elasticloadbalancing:DescribeRules",
            "elasticloadbalancing:DescribeSSLPolicies",
            "elasticloadbalancing:DescribeTags",
            "elasticloadbalancing:DescribeTargetGroups",
            "elasticloadbalancing:DescribeTargetGroupAttributes",
            "elasticloadbalancing:DescribeTargetHealth",
            "elasticloadbalancing:ModifyListener",
            "elasticloadbalancing:ModifyLoadBalancerAttributes",
            "elasticloadbalancing:ModifyRule",
            "elasticloadbalancing:ModifyTargetGroup",
            "elasticloadbalancing:ModifyTargetGroupAttributes",
            "elasticloadbalancing:RegisterTargets",
            "elasticloadbalancing:RemoveListenerCertificates",
            "elasticloadbalancing:RemoveTags",
            "elasticloadbalancing:SetIpAddressType",
            "elasticloadbalancing:SetSecurityGroups",
            "elasticloadbalancing:SetSubnets",
            "elasticloadbalancing:SetWebACL"
          ],
          "Resource": "*"
        },
        {
          "Effect": "Allow",
          "Action": [
            "iam:CreateServiceLinkedRole",
            "iam:GetServerCertificate",
            "iam:ListServerCertificates"
          ],
          "Resource": "*"
        },
        {
          "Effect": "Allow",
          "Action": [
            "cognito-idp:DescribeUserPoolClient"
          ],
          "Resource": "*"
        },
        {
          "Effect": "Allow",
          "Action": [
            "waf-regional:GetWebACLForResource",
            "waf-regional:GetWebACL",
            "waf-regional:AssociateWebACL",
            "waf-regional:DisassociateWebACL"
          ],
          "Resource": "*"
        },
        {
          "Effect": "Allow",
          "Action": [
            "tag:GetResources",
            "tag:TagResources"
          ],
          "Resource": "*"
        },
        {
          "Effect": "Allow",
          "Action": [
            "waf:GetWebACL"
          ],
          "Resource": "*"
        }
      ]
  api:
    loadBalancer:
      class: Classic
      type: Internal
  authorization:
    rbac: {}
  channel: stable
  cloudLabels:
    Business Unit: Infrastructure
    Contact: me@nowhere.com
    Service Role: k8s
    Stack: production
  cloudProvider: aws
  configBase: s3://k8s/v3/208.k8s
  dnsZone: 208.k8s
  docker:
    registryMirrors:
    - https://docker-hub.dev
    storage: overlay2,overlay,aufs
    version: 19.03.4
  encryptionConfig: true
  etcdClusters:
  - etcdMembers:
    - instanceGroup: master-eu-north-1a-1
      name: "1"
      volumeIops: 1000
      volumeSize: 20
      volumeType: io1
    - instanceGroup: master-eu-north-1a-2
      name: "2"
      volumeIops: 1000
      volumeSize: 20
      volumeType: io1
    - instanceGroup: master-eu-north-1a-3
      name: "3"
      volumeIops: 1000
      volumeSize: 20
      volumeType: io1
    manager:
      env:
      - name: ETCD_AUTO_COMPACTION_MODE
        value: revision
      - name: ETCD_AUTO_COMPACTION_RETENTION
        value: "10000"
      - name: ETCD_QUOTA_BACKEND_BYTES
        value: "4294967296"
    name: main
    version: 3.4.13
  - etcdMembers:
    - instanceGroup: master-eu-north-1a-1
      name: "1"
    - instanceGroup: master-eu-north-1a-2
      name: "2"
    - instanceGroup: master-eu-north-1a-3
      name: "3"
    name: events
    version: 3.4.13
  fileAssets:
  - content: |
      apiVersion: audit.k8s.io/v1beta1
      kind: Policy
      rules:
        # The following requests were manually identified as high-volume and low-risk,
        # so drop them.
        - level: None
          resources:
            - group: ""
              resources:
                - endpoints
                - services
                - services/status
          users:
            - 'system:kube-proxy'
          verbs:
            - watch

        - level: None
          resources:
            - group: ""
              resources:
                - nodes
                - nodes/status
          userGroups:
            - 'system:nodes'
          verbs:
            - get

        - level: None
          namespaces:
            - kube-system
          resources:
            - group: ""
              resources:
                - endpoints
          users:
            - 'system:kube-controller-manager'
            - 'system:kube-scheduler'
            - 'system:serviceaccount:kube-system:endpoint-controller'
          verbs:
            - get
            - update

        - level: None
          resources:
            - group: ""
              resources:
                - namespaces
                - namespaces/status
                - namespaces/finalize
          users:
            - 'system:apiserver'
          verbs:
            - get

        # Don't log HPA fetching metrics.
        - level: None
          resources:
            - group: metrics.k8s.io
          users:
            - 'system:kube-controller-manager'
          verbs:
            - get
            - list

        # Don't log these read-only URLs.
        - level: None
          nonResourceURLs:
            - '/healthz*'
            - /version
            - '/swagger*'

        # Don't log events requests.
        - level: None
          resources:
            - group: ""
              resources:
                - events

        # node and pod status calls from nodes are high-volume and can be large, don't log responses for expected updates from nodes
        - level: Request
          omitStages:
            - RequestReceived
          resources:
            - group: ""
              resources:
                - nodes/status
                - pods/status
          users:
            - kubelet
            - 'system:node-problem-detector'
            - 'system:serviceaccount:kube-system:node-problem-detector'
          verbs:
            - update
            - patch

        - level: Request
          omitStages:
            - RequestReceived
          resources:
            - group: ""
              resources:
                - nodes/status
                - pods/status
          userGroups:
            - 'system:nodes'
          verbs:
            - update
            - patch

        # deletecollection calls can be large, don't log responses for expected namespace deletions
        - level: Request
          omitStages:
            - RequestReceived
          users:
            - 'system:serviceaccount:kube-system:namespace-controller'
          verbs:
            - deletecollection

        # Secrets, ConfigMaps, and TokenReviews can contain sensitive & binary data,
        # so only log at the Metadata level.
        - level: Metadata
          omitStages:
            - RequestReceived
          resources:
            - group: ""
              resources:
                - secrets
                - configmaps
            - group: authentication.k8s.io
              resources:
                - tokenreviews
        # Get repsonses can be large; skip them.
        - level: Request
          omitStages:
            - RequestReceived
          resources:
            - group: ""
            - group: admissionregistration.k8s.io
            - group: apiextensions.k8s.io
            - group: apiregistration.k8s.io
            - group: apps
            - group: authentication.k8s.io
            - group: authorization.k8s.io
            - group: autoscaling
            - group: batch
            - group: certificates.k8s.io
            - group: extensions
            - group: metrics.k8s.io
            - group: networking.k8s.io
            - group: policy
            - group: rbac.authorization.k8s.io
            - group: scheduling.k8s.io
            - group: settings.k8s.io
            - group: storage.k8s.io
          verbs:
            - get
            - list
            - watch

        # Default level for known APIs
        - level: RequestResponse
          omitStages:
            - RequestReceived
          resources:
            - group: ""
            - group: admissionregistration.k8s.io
            - group: apiextensions.k8s.io
            - group: apiregistration.k8s.io
            - group: apps
            - group: authentication.k8s.io
            - group: authorization.k8s.io
            - group: autoscaling
            - group: batch
            - group: certificates.k8s.io
            - group: extensions
            - group: metrics.k8s.io
            - group: networking.k8s.io
            - group: policy
            - group: rbac.authorization.k8s.io
            - group: scheduling.k8s.io
            - group: settings.k8s.io
            - group: storage.k8s.io

        # Default level for all other requests.
        - level: Metadata
          omitStages:
            - RequestReceived
    name: kubernetes-audit
    path: /srv/kubernetes/audit.yaml
    roles:
    - Master
  hooks:
  - execContainer:
      command:
      - sh
      - -c
      - |
        chroot /rootfs apt-get update && DEBIAN_FRONTEND=noninteractive chroot /rootfs apt-get install -y --force-yes -o Dpkg::Options::="--force-confdef" acpi-support
      image: busybox
    roles:
    - Node
  - manifest: |
      Type=oneshot
      ExecStart=/sbin/sysctl net.ipv4.neigh.default.gc_thresh3=16384 ; /sbin/sysctl net.ipv4.neigh.default.gc_thresh2=8192 ; /sbin/sysctl -p
    name: increase-neigh-gc-thresh.service
  iam:
    allowContainerRegistry: true
    legacy: false
  kubeAPIServer:
    auditLogMaxAge: 10
    auditLogMaxBackups: 1
    auditLogMaxSize: 100
    auditLogPath: /var/log/kube-apiserver-audit.log
    auditPolicyFile: /srv/kubernetes/audit.yaml
    featureGates:
      TTLAfterFinished: "true"
    oidcClientID: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.xxxx.googleusercontent.com
    oidcIssuerURL: https://accounts.google.com
    oidcUsernameClaim: email
  kubeControllerManager:
    allocateNodeCIDRs: false
    featureGates:
      TTLAfterFinished: "true"
  kubeDNS:
    provider: CoreDNS
  kubeScheduler:
    featureGates:
      TTLAfterFinished: "true"
  kubelet:
    anonymousAuth: false
    authenticationTokenWebhook: true
    authorizationMode: Webhook
    cloudProvider: aws
    featureGates:
      TTLAfterFinished: "true"
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.19.10
  masterPublicName: api.208.k8s
  networkCIDR: 10.208.0.0/15
  networkID: vpc-xxxxxxxxxxxxxxxxx
  networking:
    amazonvpc:
      env:
      - name: AWS_VPC_K8S_CNI_EXTERNALSNAT
        value: "true"
      - name: WARM_IP_TARGET
        value: "1"
  nonMasqueradeCIDR: 10.208.0.0/15
  serviceClusterIPRange: 10.209.160.0/20
  subnets:
  - id: subnet-xxxxxxxxxxxxxxxxxx
    name: app
    type: Private
    zone: eu-north-1a
  - id: subnet-xxxxxxxxxxxxxxxxxx
    name: papp
    type: Private
    zone: eu-north-1a
  - id: subnet-xxxxxxxxxxxxxxxxxx
    name: sys
    type: Private
    zone: eu-north-1a
  - id: subnet-xxxxxxxxxxxxxxxxxx
    name: dbmng
    type: Private
    zone: eu-north-1a
  topology:
    dns:
      type: Private
    masters: private
    nodes: private

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2018-11-23T19:42:33Z"
  labels:
    kops.k8s.io/cluster: 208.k8s
  name: group1
spec:
  additionalSecurityGroups:
  - sg-xxxxxxxxxxxxxxxxxx
  - sg-xxxxxxxxxxxxxxxxxx
  - sg-xxxxxxxxxxxxxxxxxx
  - sg-xxxxxxxxxxxxxxxxxx
  cloudLabels:
    Contact: me@nowhere.com
    Stack: production
    k8s.io/cluster-autoscaler/208.k8s: "1"
    k8s.io/cluster-autoscaler/enabled: "1"
    kubernetes.io/cluster/208.k8s: "1"
  image: 415208756561/kops-20-04 1615540089
  machineType: c5.9xlarge
  maxPrice: "10.0"
  maxSize: 1000
  minSize: 0
  mixedInstancesPolicy:
    instances:
    - c5.9xlarge
    - c5.4xlarge
    - c5n.4xlarge
    - c5n.9xlarge
    - c5n.18xlarge
    - c5.12xlarge
    - c5.18xlarge
    - c5a.4xlarge
    - c5a.8xlarge
    - c5a.12xlarge
    - c5a.16xlarge
    - c5.xlarge
    - c5.2xlarge
    - c5a.xlarge
    - c5a.2xlarge
    - c5n.xlarge
    - c5n.2xlarge
    onDemandAboveBase: 0
    spotAllocationStrategy: capacity-optimized
    spotInstancePools: 0
  nodeLabels:
    kops.k8s.io/instancegroup: group1
    node-role.kubernetes.io/group1: group1
  role: Node
  rollingUpdate:
    maxSurge: 100%
  rootVolumeEncryption: true
  rootVolumeOptimization: true
  rootVolumeSize: 200
  rootVolumeThroughput: 250
  rootVolumeType: gp3
  subnets:
  - app
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**
Not relevant here, imho

**9. Anything else do we need to know?**
If I terminate the instances where the current "non-working" pods are running, some process just replaces more aws-node pods on existing nodes, putting me into infinite loop of manual replacing all the nodes, as `kops rolling-update cluster` is not working. I've tried to search if this behavior is described somewhere, or is there some switch which controls this behavior, but to no avail.
So my question is: is this expected behavior, or bug? If former, then what is the recommended way to change container runtime on existing cluster?

**Comment by ValeriiVozniuk** [comment]
In kubelet log on existing node I see a following entry
```
Jun 18 09:19:09 ip-10-209-22-22 kubelet[20590]: I0618 09:19:09.898659   20590 kubelet.go:1877] SyncLoop (DELETE, "api"): "aws-node-zd2k7_kube-system(d1113132-9d59-43f8-a5a0-fda2349ffc8b)"
```

**Comment by ValeriiVozniuk** [comment]
Took one more step, in kube-controller-manager found following entries
```
I0618 09:19:09.890287       1 controller_utils.go:604] "Deleting pod" controller="aws-node" pod="kube-system/aws-node-zd2k7"
I0618 09:19:09.897394       1 event.go:291] "Event occurred" object="kube-system/aws-node" kind="DaemonSet" apiVersion="apps/v1" type="Normal" reason="SuccessfulDelete" message="Deleted pod: aws-node-zd2k7"
```
But still not clear why that pod was replaced on existing nodes

**Comment by hakman** [comment]
Hi @ValeriiVozniuk. My guess is that this is similar to https://github.com/kubernetes/kops/issues/11589.
Ideally, AWS VPC CNI would just autodetect the container runtime, but at the moment ^^^ was the best solution we could think of.

**Comment by ValeriiVozniuk** [comment]
Hi @hakman,
Thank you, this link inside your link https://github.com/kubernetes/kops/pull/11590 describe my case better. Will definitely need to test it on kops 1.20

**Comment by hakman** [comment]
I don't think 1.20 with that fix is yet released, but next release should include it.

**Comment by hakman** [comment]
[kOps 1.20.2](https://github.com/kubernetes/kops/releases/tag/v1.20.2) was just released and should fix your issue. Please reopen or create a new issue if that's not the case.

**Comment by ValeriiVozniuk** [comment]
@hakman thank you, will check that version if it will be possible. For now I've solved the issue for 1.19.x with
```kubectl patch ds/aws-node -n kube-system -p '{"spec":{"updateStrategy":{"type":"OnDelete"}}}'
```
